repos/system_upgrade/el7toel8/actors/checkkdegnome/actor.py

```diff
@@ -0,0 +1,137 @@
+        with open(path, "r") as f:
+            for line in f.readlines():
+                if "xsession" in line.lower():
+                    default_xsession = line.split("=")[1].lower()
+        return default_xsession
+
+    def check_app_in_use(self, app):
+        """
+        Method return True if application was used in last month, False in other cases.
+        """
+        path = "{0}/.kde/share/config/{1}rc".format(os.environ.get("HOME"), app)
+        if os.path.exists(path) and os.path.isfile(path):
+            last_modified = os.stat(path).st_mtime
+            if last_modified >= int(time.time() - 2592000):
+                self.log.info("Application {0} is actively used".format(app))
+                return True
+        return False
+
+    def is_installed(self, app):
+        """
+        Wrapper for "rpm -q <app>" command, returns its return value, output is not supressed
+        """
+        return subprocess.call(["rpm", "-q", app])
+
+    def process(self):
+        apps_in_use = 0
+        self.log.info("  Detecting desktop environments  ")
+        self.log.info("==================================")
+
+        # Detect installed desktops by their startup files
+        kde_desktop_installed = self.is_executable("/usr/bin/startkde")
+        gnome_desktop_installed = self.is_executable("/usr/bin/gnome-session")
+        self.log.info("* KDE installed: {0}".format(kde_desktop_installed))
+        self.log.info("* Gnome installed: {0}".format(gnome_desktop_installed))
+        self.log.info("----------------------------------")
+
+        # No desktop installed, we don't even care about apps as they are most likely not used or even installed
+        if not kde_desktop_installed and not gnome_desktop_installed:
+            self.log.info("INFO: No desktop installed. Continuing with the upgrade.")
+            self.produce(CanUpgrade(canUpgrade=0))
+            return
+
+        if kde_desktop_installed:
+            self.log.info("INFO: KDE desktop is installed. Checking what we can do about it.")
+            if not gnome_desktop_installed:
+                self.log.info("ERROR: Cannot perform the upgrade because there is no other desktop installed.")
+                self.produce(CanUpgrade(canUpgrade=1))
+                return
+
+            # Assume GNOME is installed in this state
+
+            user = os.environ.get("USER")
+            default_xsession = self.get_xsession("/var/lib/AccountsService/users/{0}".format(user))
+            if not default_xsession:
+                self.log.info("WARN: Unable to detect default session.")
+            else:
+                if "plasma" in default_xsession:  # using in because there can be some white spaces
+                    self.log.info("INFO: KDE used as default session.")
+                    self.log.info("INFO: Upgrade can be performed, but KDE desktop will be removed in favor of GNOME")
+                    self.produce(CanUpgrade(canUpgrade=2))
```

Comment:
@bocekm What is the best way how to handle the state when upgrade can be performed, but KDE desktop will be removed? Just log it (and maybe set the log level to "warning") or give some stronger feedback? Thank you for the answer!

Reply:
I'd create an informative-only report message (again, using framework function `create_report`) with low or medium severity (`reporting.Severity(reporting.Severity.MEDIUM)`). With no remediation (`reporting.Remediation`) as this has no solution.

Reply:
Basically, expect that customers will not read the log messages in 99% of cases. Whatever you want the customer to know needs to land in the pre-upgrade report which we generate in both plaintext and HTML version.

Reply:
Ok, I will do that this way. Thanks!